File: julia/math/modpow.jl
# Print modular powers of fixed triples, each squared and halved by repeated squaring.
# Run: julia modpow.jl

const CASES = [(2, 1000, 1000003), (3, 200, 50), (5, 117, 19), (10, 18, 9999991)]

function modpow(base, exponent, modulus)
    result = 1
    base %= modulus
    while exponent > 0
        if isodd(exponent)
            result = result * base % modulus
        end
        base = base * base % modulus
        exponent ÷= 2
    end
    return result
end

for (base, exponent, modulus) in CASES
    println(join((base, exponent, modulus, modpow(base, exponent, modulus)), " "))
end

File: clojure/math/modpow.clj
;; Print modular powers of fixed triples, each squared and halved by repeated squaring.
;; Run: clojure -M modpow.clj

(require '[clojure.string :as string])

(def cases [[2 1000 1000003] [3 200 50] [5 117 19] [10 18 9999991]])

(defn modpow [base exponent modulus]
  (loop [base (mod base modulus) exponent exponent result 1]
    (if (zero? exponent)
      result
      (recur (mod (* base base) modulus)
             (quot exponent 2)
             (if (odd? exponent) (mod (* result base) modulus) result)))))

(doseq [[base exponent modulus] cases]
  (println (string/join " " [base exponent modulus (modpow base exponent modulus)])))
